src/nft/components/details/detailsV2/BuyButton.tsx

```diff
@@ -24,30 +87,136 @@ const Price = styled.div`
+  const { account } = useWeb3React()
+  const [accountDrawerOpen, toggleWalletDrawer] = useAccountDrawer()
   const { fetchAndPurchaseSingleAsset, isLoading: isLoadingRoute } = useBuyAssetCallback()
+  const [connectingToWallet, setConnectingToWallet] = useState(false)
+  const [addToBagExpanded, setAddToBagExpanded] = useState(false)
+  const navigate = useNavigate()
+  const setSellPageState = useProfilePageState((state) => state.setProfilePageState)
+  const resetSellAssets = useSellAsset((state) => state.reset)
+  const clearCollectionFilters = useWalletCollections((state) => state.clearCollectionFilters)
+
+  const { walletAssets } = useNftBalance(account ?? '', [], [{ address: asset.address, tokenId: asset.tokenId }], 1)
+  const walletAsset = walletAssets?.[0]
+
   const price = asset.sellorders?.[0]?.price.value
+  const ownsAsset = Boolean(walletAsset) && account?.toLowerCase() === asset.ownerAddress?.toLowerCase()
+
+  const assetInBag = useIsAssetInBag(asset)
+  const { addAssetsToBag, removeAssetsFromBag } = useBag(
+    ({ addAssetsToBag, removeAssetsFromBag }) => ({
+      addAssetsToBag,
+      removeAssetsFromBag,
+    }),
+    shallow
+  )
+
+  const navigateToProfile = useCallback(() => {
+    resetSellAssets()
+    setSellPageState(ProfilePageStateType.VIEWING)
+    clearCollectionFilters()
+    navigate('/nfts/profile')
+  }, [clearCollectionFilters, navigate, resetSellAssets, setSellPageState])
+
+  const oneClickBuyAsset = useCallback(() => {
+    if (!account) {
+      if (!accountDrawerOpen) {
+        toggleWalletDrawer()
+      }
+      setConnectingToWallet(true)
+      setTimeout(() => setConnectingToWallet(false), 20000)
+    } else {
+      fetchAndPurchaseSingleAsset(asset)
+    }
+  }, [account, accountDrawerOpen, asset, fetchAndPurchaseSingleAsset, toggleWalletDrawer])
+
+  useEffect(() => {
+    if (connectingToWallet && account) {
+      setConnectingToWallet(false)
+      fetchAndPurchaseSingleAsset(asset)
+    }
+  }, [connectingToWallet, account, fetchAndPurchaseSingleAsset, asset])
+
+  if (ownsAsset) {
+    return (
+      <ButtonContainer dataPage={Boolean(dataPage)}>
+        <StyledBuyButton shouldHide={false} onClick={navigateToProfile}>
+          {price ? (
+            <>
+              <Trans>Edit Listing</Trans>
+              <Price>{formatNumber(price)} ETH</Price>
+            </>
+          ) : (
+            <Trans>List</Trans>
+          )}
+        </StyledBuyButton>
+      </ButtonContainer>
+    )
+  }
 
   if (!price) {
-    return <OfferButton />
+    if (dataPage) {
+      return null
+    }
+
+    return (
+      <NotAvailableButton disabled>
+        <Trans>Not available for purchase</Trans>
+      </NotAvailableButton>
+    )
+  }
+
+  const secondaryButtonCta = assetInBag ? t`Remove from Bag` : t`Add to Bag`
+  const SecondaryButtonIcon = () =>
+    assetInBag ? <CondensedBagIcon width="24px" height="24px" /> : <AddToBagIcon width="24px" height="24px" />
```

Comment:
More of an ask for design but we should have a specific icon for removing from bag as opposed to just using the generic bag icon. This is very noticeable on mobile where you don't get the hover text,
![Screen Shot 2023-05-31 at 15 40 56 ](https://github.com/Uniswap/interface/assets/11512321/b7bb51d0-d768-4ac2-a252-bab599b39512)